BLUESPAWN-client/src/util/configurations/RegistryKey.cpp

```diff
@@ -0,0 +1,491 @@
+			SetLastError(status);
+			return { nullptr, 0 };
+		}
+
+		auto lpbValue = new BYTE[dwDataSize];
+		status = RegQueryValueExW(hkBackingKey, ValueName.c_str(), 0, nullptr, lpbValue, &dwDataSize);
+		if(status != ERROR_SUCCESS){
+			SetLastError(status);
+			return { nullptr, 0 };
+		}
+
+		return { lpbValue, dwDataSize };
+	}
+
+	std::optional<RegistryType> RegistryKey::GetValueType(const std::wstring& ValueName) const {
+		if(!Exists()){
+			SetLastError(FILE_DOES_NOT_EXIST);
+			return std::nullopt;
+		}
+
+		DWORD dwType{};
+
+		LSTATUS status = RegQueryValueExW(hkBackingKey, ValueName.c_str(), 0, &dwType, nullptr, nullptr);
+		if(status != ERROR_SUCCESS && status != ERROR_MORE_DATA){
+			SetLastError(status);
+			return std::nullopt;
+		}
+
+		if(dwType == REG_SZ){
+			return RegistryType::REG_SZ_T;
+		} else if(dwType == REG_EXPAND_SZ){
+			return RegistryType::REG_EXPAND_SZ_T;
+		} else if(dwType == REG_MULTI_SZ){
+			return RegistryType::REG_MULTI_SZ_T;
+		} else if(dwType == REG_DWORD){
+			return RegistryType::REG_DWORD_T;
+		}
+
+		return RegistryType::REG_BINARY_T;
+	}
+
+	template<class T>
+	std::optional<T> RegistryKey::GetValue(const std::wstring& wsValueName) const {
+		if(ValueExists(wsValueName)){
+			return GetRawValue(wsValueName).Dereference<T>();
+		}
+		return std::nullopt;
+	}
+
+	template std::optional<DWORD> RegistryKey::GetValue(const std::wstring& wsValueName) const;
+
+	template<>
+	std::optional<std::wstring> RegistryKey::GetValue(const std::wstring& wsValueName) const {
+		if(ValueExists(wsValueName)){
+			return GetRawValue(wsValueName).ReadWString();
+		}
+		return std::nullopt;
+	}
+
+	template<>
+	std::optional<std::vector<std::wstring>> RegistryKey::GetValue(const std::wstring& wsValueName) const {
+		if(ValueExists(wsValueName)){
+			std::vector<std::wstring> strings{};
+
+			LPCWSTR data = reinterpret_cast<LPCWSTR>((LPVOID) GetRawValue(wsValueName));
+
+			while(*data){
+				std::wstring str = data;
+				strings.emplace_back(data);
+
+				data += str.length() + 1;
+			}
+
+			return strings;
+		}
+
+		return std::nullopt;
+	}
+
+	bool RegistryKey::SetRawValue(const std::wstring& name, AllocationWrapper bytes, DWORD dwType) const {
+		if(!Exists()){
+			return false;
+		}
+
+		LSTATUS status = RegSetValueEx(hkBackingKey, name.c_str(), 0, dwType, reinterpret_cast<BYTE*>((LPVOID) bytes), bytes.GetSize());
+		if(status != ERROR_SUCCESS){
+			SetLastError(status);
+			return false;
+		}
+
+		return true;
+	}
+
+	template<class T>
+	bool RegistryKey::SetValue(const std::wstring& name, T value, DWORD size, DWORD type) const {
+		return SetRawValue(name, { reinterpret_cast<BYTE*>(value), size, AllocationWrapper::STACK_ALLOC }, type);
+	}
+
+	template<>
+	bool RegistryKey::SetValue(const std::wstring& name, LPCWSTR value, DWORD size, DWORD type) const {
+		return RegistryKey::SetRawValue(name, { PBYTE(value), wcslen(value), AllocationWrapper::STACK_ALLOC }, type);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=4967854)